### PR TITLE
NormalizingFlow class and Affine transformation created using Pytorch

### PR DESCRIPTION
**Pull request by JoseAntonioSiguenza**
## Creation of NormalizingFlow class using Pytorch.

Fix one model of #2863 

The main classes of this PR are `NormalizingFlow` (model) and `Affine` (transformation) using Pytorch. Documentation and tests need to be done, nevertheless, I wanted to show some progress. Also, work needs to be done with the `NormalizingFlowModel` class. This class makes it possible to work following DeepChem API by using a `TorchModel`.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend discussing your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by JoseAntonioSiguenza** [comment]
I've split the PR and this version contains the `Affine` transformation layer.  This commit also includes docstrings, type hints, and a test.

**Comment by JoseAntonioSiguenza** [comment]
> I've split the PR and this version contains the `Affine` transformation layer. This commit also includes docstrings, type hints, and a test.

CC @paupaiz @ARY2260

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+class Affine(nn.Module):
+  """Class which performs the Affine transformation.
+
+  This transformation is based on the affinity of the base distribution with
+  the target distribution. A geometric transformation is applied where
+  the parameters performs changes on the scale and shift of a function
+  (inputs).
+  
+  Normalizing Flow transformations must be bijective in order to compute
+  the logartihm of jacobian's determinat. For this reason, transformations
```
Comment (ARY2260) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:32]: typo : determinat 
corrected : determinant

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        Value of the Nth dimenssion of the dataset. 
+
+      """
+
+      super().__init__()
+      self.dim = dim
+      self.scale = nn.Parameter(torch.zeros(self.dim))
+      self.shift = nn.Parameter(torch.zeros(self.dim))
+
+  def forward(self, x):
```
Comment (ARY2260) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:52]: I think type annotation and docstring missing for forward() and inverse().

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+
+import deepchem as dc
+from deepchem.models.losses import Loss
+from deepchem.models.models import Model
+from deepchem.models.torch_models import TorchModel
+from deepchem.models.optimizers import Optimizer, Adam
+from deepchem.utils.typing import OneOrMany
+from deepchem.utils.data_utils import load_from_disk, save_to_disk
```
Comment (ARY2260) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:19]: Flake8 tests show errors for unused imports. I am not sure if it is tested in checks performed on PRs. Though if it does happen, your PR might get stuck in checks.

**Comment by JoseAntonioSiguenza** [comment]
Thanks a lot for the feedback @ARY2260!! Now I've run the code according to conventions and CI, and still working on the `forward `and `inverse` docstrings.

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  """Class which performs the Affine transformation.
+
+  This transformation is based on the affinity of the base distribution with
+  the target distribution. A geometric transformation is applied where
+  the parameters performs changes on the scale and shift of a function
+  (inputs).
+
+  Normalizing Flow transformations must be bijective in order to compute
+  the logartihm of jacobian's determinant. For this reason, transformations
+  must perform a forward and inverse pass.
```
Comment (rbharath) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:24]: Can you add a usage example?

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import logging
+
+import torch
+import torch.nn as nn
+
+
+logger = logging.getLogger(__name__)
+
+
+class Affine(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:14]: Can you add to docs at https://github.com/deepchem/deepchem/blob/master/docs/source/api_reference/layers.rst?

File: deepchem/models/tests/test_normalizing_flow_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  from torch.distributions import MultivariateNormal
+  from deepchem.models.torch_models.normalizing_flows_pytorch import Affine
+  has_torch = True
+except:
+  has_torch = False
+
+
+@unittest.skipIf(not has_torch, 'torch is not installed')
+@pytest.mark.torch
+def test_Affine():
```
Comment (rbharath) [inline deepchem/models/tests/test_normalizing_flow_pytorch.py:20]: Can you add a docstring?

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  """Class which performs the Affine transformation.
+
```
Comment (rbharath) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:16]: What are shapes in, what are shape out? Can you add to docstring

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  (inputs).
+
+  Normalizing Flow transformations must be bijective in order to compute
+  the logarithm of jacobian's determinant. For this reason, transformations
+  must perform a forward and inverse pass.
+
+  Example
+  --------
+  >>>
+  >> import deepchem as dc
```
Comment (rbharath) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:29]: All these should be `>>>` since otherwise test will not run. (Some examples leave `>>>` out because the example is known to not run on doctest)

File: deepchem/models/torch_models/normalizing_flows_pytorch.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    y = torch.exp(self.scale) * x + self.shift
+    det_jacobian = torch.exp(self.scale.sum())
+    log_det_jacobian = torch.ones(y.shape[0]) * torch.log(det_jacobian)
+
+    return y, log_det_jacobian
+
+  def inverse(self, y: Sequence) -> Tuple[torch.Tensor, torch.Tensor]:
+    """
+    Performs a transformation between two different distributions.
```
Comment (rbharath) [inline deepchem/models/torch_models/normalizing_flows_pytorch.py:77]: Needs to follow numpydoc style
